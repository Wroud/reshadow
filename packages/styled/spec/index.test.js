/**
 * @jest-environment enzyme
 */

import React from 'react';
import {render, shallow} from 'enzyme';

const getStyles = () =>
    [...document.getElementsByTagName('style')]
        .map(x => x.textContent)
        .join('');

let styled, ThemeProvider, css, createGlobalStyle;

describe('styled', () => {
    beforeEach(() => {
        [...document.getElementsByTagName('style')].forEach(x => {
            x.remove();
        });

        jest.isolateModules(() => {
            ({
                default: styled,
                ThemeProvider,
                css,
                createGlobalStyle,
            } = require('..'));
        });
    });

    it('should apply styles', () => {
        const Button = styled.button`
            color: red;
        `;

        const wrapper = render(<Button>click me</Button>);
        expect(wrapper).toMatchSnapshot();
        expect(getStyles()).toMatchSnapshot();
    });

    it('should apply nested styles', () => {
        const Button = styled.button`
            color: red;

            & span {
                margin-right: 10px;
            }
        `;

        const wrapper = render(<Button>click me</Button>);
        expect(wrapper).toMatchSnapshot();
        expect(getStyles()).toMatchSnapshot();
    });

    it('should apply styles for Component', () => {
        const Button = styled(props => <button {...props} />)`
            color: red;
        `;

        const wrapper = render(<Button>click me</Button>);
        expect(wrapper).toMatchSnapshot();
        expect(getStyles()).toMatchSnapshot();
    });

    it('should apply dynamic styles', () => {
        const Button = styled.button`
            color: ${props => props.color};
        `;

        const wrapper = render(<Button color="red">click me</Button>);
        expect(wrapper).toMatchSnapshot();
        expect(getStyles()).toMatchSnapshot();
    });

    it('should work with attrs', () => {
        const Button = styled.button.attrs({
            type: 'submit',
            name: props => props.color,
        })`
            color: ${props => props.color};
        `;

        const wrapper = render(<Button color="red">click me</Button>);
        expect(wrapper).toMatchSnapshot();
    });

    it('should work with as', () => {
        const Button = styled.button`
            color: ${props => props.color};
        `;

        const wrapper = render(
            <Button as="a" href="#" color="red">
                click me
            </Button>,
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('should work with theme', () => {
        const Button = styled.button`
            color: ${props => props.theme.color};
        `;

        const theme = {
            color: 'red',
        };

        const wrapper = render(
            <ThemeProvider theme={theme}>
                <Button>click me</Button>
            </ThemeProvider>,
        );
        expect(wrapper).toMatchSnapshot();
        expect(getStyles()).toMatchSnapshot();
    });

    it('should work with css prop', () => {
        const Button = styled.button`
            color: ${props => props.color};
        `;

        const wrapper = render(
            <Button
                color="red"
                css={`
                    padding: 5px 10px;
                `}
            >
                click me
            </Button>,
        );

        expect(wrapper).toMatchSnapshot();
        expect(getStyles()).toMatchSnapshot();
    });

    it('should work with css prop function', () => {
        const Button = styled.button`
            color: ${props => props.color};
        `;

        const wrapper = render(
            <Button
                color="red"
                top={5}
                css={props => `
                    padding: ${props.top}px 10px;
                `}
            >
                click me
            </Button>,
        );

        expect(wrapper).toMatchSnapshot();
        expect(getStyles()).toMatchSnapshot();
    });

    it('should work with css prop with tagged literal', () => {
        const Button = styled.button`
            color: ${props => props.color};
        `;

        const wrapper = render(
            <Button
                color="red"
                top={5}
                css={css`
                    padding: 5px 10px;
                `}
            >
                click me
            </Button>,
        );

        expect(wrapper).toMatchSnapshot();
        expect(getStyles()).toMatchSnapshot();
    });

    it('should work with mixins', () => {
        const padding = css`
            padding: 10px;
        `;

        const after = css`
            &::after {
                content: '';
            }
        `;

        const before = css`
            ::before {
                content: '';
            }
        `;

        const Button = styled.button`
            ${padding}
            ${after}
            ${before}
            color: ${props => props.color};
        `;

        const wrapper = render(<Button color="red">click me</Button>);

        expect(wrapper).toMatchSnapshot();
        expect(getStyles()).toMatchSnapshot();
    });

    it('should work with nested mixins', () => {
        const complexMixin = css`
            color: ${props => (props.whiteColor ? 'white' : 'black')};
        `;

        const Button = styled.button`
            ${props => (props.complex ? complexMixin : 'color: blue;')};
        `;

        const wrapper = shallow(<Button complex>click me</Button>);

        expect(wrapper.render()).toMatchSnapshot();
        expect(getStyles()).toMatchSnapshot();

        wrapper.setProps({whiteColor: true});
        expect(wrapper.render()).toMatchSnapshot();
        expect(getStyles()).toMatchSnapshot();

        wrapper.setProps({complex: false});
        expect(wrapper.render()).toMatchSnapshot();
        expect(getStyles()).toMatchSnapshot();
    });

    it('should create global style', () => {
        const GlobalStyle = createGlobalStyle`
            body {
                color: ${props => (props.whiteColor ? 'white' : 'black')};
            }
        `;

        const wrapper = shallow(<GlobalStyle whiteColor />);

        expect(wrapper.render()).toMatchSnapshot();
        expect(getStyles()).toMatchSnapshot();
    });
});
