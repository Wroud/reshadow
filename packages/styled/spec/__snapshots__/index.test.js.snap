// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`styled should apply dynamic styles 1`] = `
<button
  class="_sna6ae"
  style="--1v6vp2v_1:red"
>
  click me
</button>
`;

exports[`styled should apply dynamic styles 2`] = `"._sna6ae{color:var(--1v6vp2v_1);}"`;

exports[`styled should apply nested styles 1`] = `
<button
  class="_5jpyup"
>
  click me
</button>
`;

exports[`styled should apply nested styles 2`] = `"._5jpyup span{margin-right:10px;}._5jpyup{color:red;}"`;

exports[`styled should apply styles 1`] = `
<button
  class="_1if0x5c"
>
  click me
</button>
`;

exports[`styled should apply styles 2`] = `"._1if0x5c{color:red;}"`;

exports[`styled should apply styles for Component 1`] = `
<button
  class="_1if0x5c"
>
  click me
</button>
`;

exports[`styled should apply styles for Component 2`] = `"._1if0x5c{color:red;}"`;

exports[`styled should create global style 1`] = `null`;

exports[`styled should create global style 2`] = `"body{color:var(--10pagwf_1);}"`;

exports[`styled should deal with specificity 1`] = `
<h4
  class="_12w6gee"
  style="--egcu4c_2:20;--egcu4c_4:20"
>
  title
</h4>
`;

exports[`styled should deal with specificity 2`] = `"._12w6gee:after{-webkit-flex:1;-ms-flex:1;flex:1;-webkit-flex:var(--egcu4c_2);-ms-flex:var(--egcu4c_2);flex:var(--egcu4c_2);}._12w6gee:before{-webkit-flex:1;-ms-flex:1;flex:1;-webkit-flex:var(--egcu4c_4);-ms-flex:var(--egcu4c_4);flex:var(--egcu4c_4);}"`;

exports[`styled should process namespaced attributes and dont pass them 1`] = `
<button
  class="_zgczid _use--disabled_zgczid _use--size_m_zgczid"
>
  click me
</button>
`;

exports[`styled should process namespaced attributes and dont pass them 2`] = `"._zgczid._use--size_m_zgczid{font-size:14px;}._zgczid._use--size_s_zgczid{font-size:14px;}._zgczid._use--disabled_zgczid{opacity:0.5;}._zgczid{color:red;}"`;

exports[`styled should process namespaced attributes and dont pass them 3`] = `
<button
  class="_zgczid _use--disabled_zgczid _use--size_s_zgczid"
>
  click me
</button>
`;

exports[`styled should process namespaced attributes and dont pass them 4`] = `"._zgczid._use--size_m_zgczid{font-size:14px;}._zgczid._use--size_s_zgczid{font-size:14px;}._zgczid._use--disabled_zgczid{opacity:0.5;}._zgczid{color:red;}"`;

exports[`styled should process root props 1`] = `
<button
  class="_zdj1v2 _disabled_zdj1v2"
  disabled=""
>
  click me
</button>
`;

exports[`styled should process root props 2`] = `"._zdj1v2._disabled_zdj1v2 + [disabled] + ._zdj1v2._disabled_zdj1v2{opacity:0.5;}"`;

exports[`styled should work with as 1`] = `
<a
  class="_sna6ae"
  href="#"
  style="--1v6vp2v_1:red"
>
  click me
</a>
`;

exports[`styled should work with attrs 1`] = `
<button
  class="_sna6ae"
  name="red"
  style="--1v6vp2v_1:red"
  type="submit"
>
  click me
</button>
`;

exports[`styled should work with css prop 1`] = `
<button
  class="_1eg04s3"
  style="--1v6vp2v_1:red"
>
  click me
</button>
`;

exports[`styled should work with css prop 2`] = `"._1eg04s3{padding:5px 10px;color:var(--1v6vp2v_1);}"`;

exports[`styled should work with css prop function 1`] = `
<button
  class="_1eg04s3"
  style="--1v6vp2v_1:red"
>
  click me
</button>
`;

exports[`styled should work with css prop function 2`] = `"._1eg04s3{padding:5px 10px;color:var(--1v6vp2v_1);}"`;

exports[`styled should work with css prop with tagged literal 1`] = `
<button
  class="_1eg04s3"
  style="--1v6vp2v_1:red"
>
  click me
</button>
`;

exports[`styled should work with css prop with tagged literal 2`] = `"._1eg04s3{padding:5px 10px;color:var(--1v6vp2v_1);}"`;

exports[`styled should work with dynamic mixins 1`] = `
<button
  class="_miwxjj"
  style="--auziqq_1:-0px;--auziqq_2:0px"
  width="0"
>
  click me
</button>
`;

exports[`styled should work with dynamic mixins 2`] = `"._miwxjj{-webkit-transform:translateX(var(--auziqq_1));-ms-transform:translateX(var(--auziqq_1));transform:translateX(var(--auziqq_1));margin:var(--auziqq_2);}"`;

exports[`styled should work with dynamic mixins 3`] = `
<button
  class="_miwxjj"
  style="--auziqq_1:-200px;--auziqq_2:100px"
  width="100"
>
  click me
</button>
`;

exports[`styled should work with dynamic mixins 4`] = `"._miwxjj{-webkit-transform:translateX(var(--auziqq_1));-ms-transform:translateX(var(--auziqq_1));transform:translateX(var(--auziqq_1));margin:var(--auziqq_2);}"`;

exports[`styled should work with mixins 1`] = `
<button
  class="_wopuxb"
  style="--1txtmwt_4:red"
>
  click me
</button>
`;

exports[`styled should work with mixins 2`] = `"._wopuxb{padding:10px;color:var(--1txtmwt_4);}._wopuxb::after{content:'';}._wopuxb::before{content:'';}"`;

exports[`styled should work with nested mixins 1`] = `
<button
  class="_4ykgti"
  style="--pjcj95_1:black"
>
  click me
</button>
`;

exports[`styled should work with nested mixins 2`] = `"._4ykgti{color:var(--pjcj95_1);}"`;

exports[`styled should work with nested mixins 3`] = `
<button
  class="_4ykgti"
  style="--pjcj95_1:white"
>
  click me
</button>
`;

exports[`styled should work with nested mixins 4`] = `"._4ykgti{color:var(--pjcj95_1);}"`;

exports[`styled should work with nested mixins 5`] = `
<button
  class="_g1uhfs"
>
  click me
</button>
`;

exports[`styled should work with nested mixins 6`] = `"._4ykgti{color:var(--pjcj95_1);}._g1uhfs{color:blue;}"`;

exports[`styled should work with string mixins 1`] = `
<button
  class="_zip2hf"
  width="100"
>
  click me
</button>
`;

exports[`styled should work with string mixins 2`] = `"._zip2hf{padding:10px;}"`;

exports[`styled should work with theme 1`] = `
<button
  class="_sna6ae"
  style="--1v6vp2v_1:red"
>
  click me
</button>
`;

exports[`styled should work with theme 2`] = `"._sna6ae{color:var(--1v6vp2v_1);}"`;

exports[`styled supports function interface 1`] = `
<button
  class="_51xj47 _use--color_1ntyugu"
  style="--ajx7i6_1:red"
>
  click me
</button>
`;

exports[`styled supports function interface 2`] = `"._51xj47._use--color_1ntyugu{color:var(--ajx7i6_1);}"`;
